Домашняя работа 8

khitao opened the pull request:
Добавил в проект зависимость spring-boot-starter-security и реализовал Basic Authentication

Добавил регистрацию пользователей в приложении и организовал хранение логинов/паролей в БД.

Пользователь с ролью ADMIN создается в скрипте миграции.

Тесты минус


Тесты не получились( даже к существующим добавил @WithMockUser(roles = "ADMIN") и там где  POST запросы тесты ломаются на get все хорошо (может что-то не вижу или не понимаю)(((

khitao committed:
Домашняя работа 7

Добавил в проект зависимость spring-boot-starter-security и реализовал Basic Authentication

Добавил регистрацию пользователей в приложении и организовал хранение логинов/паролей в БД.

Пользователь с ролью ADMIN создается в скрипте миграции.

Тесты минус

madoed commented on SpringBootApp/src/main/java/ru/khodov/springbootapp/config/SecurityConfig.java:
```diff
@@ -0,0 +1,49 @@
+    @Bean
+    protected SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
+        return http
+                .csrf(AbstractHttpConfigurer::disable)
+                .cors(AbstractHttpConfigurer::disable)
```
Какие аргументы в пользу отключения CSRF и CORS?

khitao replied:
использую только localhost поэтому, как мне кажется ограничения CORS и CSRF не нужны, так как атаки CSRF в основном угрожают приложениям, доступным в Интернете + нет важных данных + проще тестировать

madoed commented on SpringBootApp/src/main/java/ru/khodov/springbootapp/controllers/WeatherApiController.java:
```diff
@@ -20,6 +21,7 @@ public WeatherApiController(WeatherApiService weatherApiClient,
     }
 
     @GetMapping("/{location}")
+    @PreAuthorize("hasAnyRole('USER', 'ADMIN')")
```
Чтобы использовать эту аннотацию нужно включить её поддержку, например через `@EnableMethodSecurity`

madoed replied:
https://docs.spring.io/spring-security/reference/servlet/authorization/method-security.html

madoed reviewed (changes requested):
Доделать и ответить на комментарии.
Включение Method security функционала, скорее всего поможет и починить тесты.

khitao committed:
Исправления Домашняя работа 8

Добавил аннотацию @EnableMethodSecurity и Тесты-mОтключил Security от прошлых тестов при помощи addFilters = false

madoed commented on SpringBootApp/src/test/java/ru/khodov/springbootapp/controllers/security/AuthenticationControllerTest.java:
```diff
@@ -0,0 +1,53 @@
+
+@SpringBootTest
+@AutoConfigureMockMvc(printOnlyOnFailure = false)
+public class AuthenticationControllerTest {
```
Тест не рабочий. Нет поднятия тестовой БД

Ниже используется UserRepository, который требует наличия подключения к БД

madoed replied:
Есть отдельный у тебя класс 

SpringBootAppApplicationTests - там объявление бд - это надо использовать здесь (через наследование)

madoed replied:
Ну и остальные тесты требующие БД (а это все, которые поднимают весь контекст через `@SpringBootTest`) по той же причине падают

khitao replied:
Они у меня почему-то запустились 🤔 наверное, поэтому я и не подумал, что нужна бд 
Хотя думал, что странно почему запускаются

khitao replied:
<img width="1417" alt="Снимок экрана 2023-11-13 в 13 36 51" src="https://github.com/khitao/TinkoffFirstHW/assets/108884890/bb154733-96f9-4a76-ab4b-1c4bc79b92b6">

khitao replied:
Очень странно, почему так могло получиться 🤔

khitao replied:
<img width="1417" alt="Снимок экрана 2023-11-13 в 17 05 01" src="https://github.com/khitao/TinkoffFirstHW/assets/108884890/e9807c8b-8890-4391-bf15-73ab06d80a78">
После исправлений

madoed replied:
А покажи свои application.properties вижу они отличны, от того что в гитхабе

Есть предположение, что ты там указываешь бд (например h2) и верный api ключ. Тесты должны работать на том application.properties, который в репозитории.

Если надо изменить, то добавляешь в test application-test.properties и подвязываешь к тестам (https://www.baeldung.com/spring-tests-override-properties)

khitao replied:
Там просто api ключ

khitao replied:
![Uploading Снимок экрана 2023-11-15 в 11.17.13.png…]()

khitao replied:
<img width="1437" alt="Снимок экрана 2023-11-15 в 11 17 53" src="https://github.com/khitao/TinkoffFirstHW/assets/108884890/af65b130-9a2a-4028-8d29-ed124a3389b0">

khitao replied:
Я что-то почитал и почти везде пишут, что единственный способ передать ключ(это передать лично) На гит хаб нельзя же пушить

madoed reviewed (changes requested):
Нужно переделать и починить тесты

khitao committed:
Исправления тестов Домашняя работа 8

khitao commented on SpringBootApp/src/test/java/ru/khodov/springbootapp/controllers/security/WeatherControllerTest.java:
Не наследовал от SpringBootAppApplicationTests тк не используется бд

madoed replied:
Используется, у тебя `@SpringBootTest` поднимает весь контекст и ниже используется 
```java

    @SpyBean
    private WeatherService weatherService;
```

Который использует реализацию сервиса без моков

khitao replied:
Там в самом сервисе в мапу идет сохранение ( я это имел ввиду)

khitao commented on SpringBootApp/gradle.properties:
Добавил тк почему-то при повторном запуске тестов в BUILD файле возникал конфликт (дублировались классы тестов) после добавления ошибка исчезла

khitao committed:
Доработки дз8

khitao committed:
Исправление дз8

madoed commented:
Все заработало кроме 
getCurrentWeather_Test_status_isOk_ADMIN
getCurrentWeather_Test_status_isOk_USER

Сюда как раз надо добавить моку для внешней интеграции

khitao committed:
Исправления тестов с API
